elastic/kibana
# [Index Management] Add data retention support to index template

Closes: https://github.com/elastic/kibana/issues/165139

## Summary
This PR adds support for configuring data retention in index template. This is only available for non legacy index templates and only when the index template creates `data streams` instead of indices.

##### How to test
1. Start elasticsearch with `yarn es snapshot --license=trial -E path.repo=/tmp/snapshots` and kibana with `yarn start`
2. Navigate to index management, click on index template tab  and create a new one
3. Verify that:
  * When selecting `Create a data stream` toggle, the section for setting up data retention shows up
  * Data retention can be configured and has a correct validation
  * Review template step should show the lifecycle values correctly in the preview and request tabs
  * Upon creating the template the details flyout should show the correct lifecycle value
  * Editing a template should initialize the values correctly


<details>
<summary>Screenshots</summary>

![Screenshot 2023-11-06 at 09 55 10](https://github.com/elastic/kibana/assets/1191206/7d3dab74-5005-481e-ad4d-a81d0589fd21)
![Screenshot 2023-11-06 at 09 55 15](https://github.com/elastic/kibana/assets/1191206/a689fb4a-466f-4403-be79-37ceec6b0bdc)
![Screenshot 2023-11-06 at 09 55 21](https://github.com/elastic/kibana/assets/1191206/43b907d7-7bce-4766-a22c-2e5469e3a7b5)
![Screenshot 2023-11-06 at 09 55 32](https://github.com/elastic/kibana/assets/1191206/1d282490-7f82-4a02-bd45-691a106f0d82)
![Screenshot 2023-11-06 at 09 55 37](https://github.com/elastic/kibana/assets/1191206/12e3ea29-793c-45de-a41f-007c2abd43bb)
![Screenshot 2023-11-06 at 09 55 54](https://github.com/elastic/kibana/assets/1191206/f6d2d51a-ae9a-4116-b013-ca765cd917b3)
![Screenshot 2023-11-06 at 09 55 58](https://github.com/elastic/kibana/assets/1191206/db84e5be-bf18-4038-ad93-f00ab9f0f74c)
![Screenshot 2023-11-06 at 09 56 04](https://github.com/elastic/kibana/assets/1191206/164f9c36-a2a7-432d-80da-1450dc05d7c1)
![Screenshot 2023-11-06 at 09 56 34](https://github.com/elastic/kibana/assets/1191206/4acf5fb1-e95b-4a82-a038-340a00bd62ad)


</details>
